File: Python/wrd_client.py
# Author D.L
# Last update at May 29, 2017
import requests


class WRDClient():
    def __init__(self, wrdUrl, username, password):
        self.logon_url = "/Account/LogOn"
        self.location_list_url = "/API/WaterQuality/Location/List"
        self.location_detail_url = "/API/WaterQuality/Location/Detail/"
        self.guideline_list_url = "/API/WaterQualityGuideline"
        self.guideline_detail_url = "/API/WaterQualityGuideline/Detail"
        self.graph_data_url = "/API/WaterQuality/Graph"
        self.root_url = wrdUrl

        headers = {'User-Agent': 'Mozilla/5.0'}
        payload = {'Username': username, 'Password': password}

        self.session = requests.session()
        r = self.session.post(self.root_url + self.logon_url, headers=headers, data=payload)
        if 200 == r.status_code:
            return
        else:
            print("Authentication session fail with status code:" + r.status_code)
            raise RuntimeError("Get location fail with status code" + r.status_code)

    def __del__(self):        
        self.session.close()

    def get_locations(self):
        response = self.__httpGet(self.root_url + self.location_list_url)
        data = list(map(lambda x: {
            "Name": x["LocationIdentifier"],
            "Description": x["LocationDescription"],
            "Latitude": x["Latitude"],
            "Longtitude": x["Longtitude"]}, response))
        return data

    def get_locations_detail(self, locationIds):
        response = self.__httpGet(self.root_url + self.location_detail_url + locationIds)
        data = {
            "Location": response["Locations"][0]["Name"],
            "Analytes": list(map(lambda x:
                                 {"Analyte": x["Analyte"],
                                  "FirstRecordDate": x["FirstRecordDate"],
                                  "LastRecordDate": x["LastRecordDate"],
                                  "NumberOfRecords": x["NumberOfRecords"]}, response["WaterQualityDatasets"]))
        }
        return data

    def get_guidelines(self):
        response = self.__httpGet(self.root_url + self.guideline_list_url)
        data = list(map(lambda x: {
            "Id": x["Id"],
            "GuidelineName": x["GuidelineName"],
            "GuidelineLongName": x["GuidelineLongName"]}, response))
        return data

    def get_guidelines_detail(self):
        return self.__httpGet(self.root_url + self.guideline_detail_url)

    def get_graph_data(self, start_date, end_date, station_name, analytes, guidelines):
        payload = {
            "StartDate": start_date,
            "EndDate": end_date,
            "IncludeDischargeData": False,
            "StationName": station_name,
            "WaterQualityStandards": list(map(lambda x: {"GuidelineName": x}, guidelines)),
            "WaterQualityDatasets": list(map(lambda x: {"Analyte": x, "LocationName": station_name}, analytes))
        }
        http_data = self.__httpPost(url=(self.root_url + self.graph_data_url), json_data=payload)

        data = {"Location": station_name,
                "AnalytesData": list(map(lambda x: {
                    "Location": x["StationId"],
                    "Analyte": x["AnalyteName"],
                    "Unit": x["Unit"],
                    "Points": list(map(lambda p: {
                        "DateTime": p["DateTime"],
                        "Value": p["Value"],
                        "DetectionLimit": p["DetectionLimit"]}, x["Points"]))}, http_data["AnalytesData"])),
                "GuidelineData": list(map(lambda x: {
                    "GuidelineName": x["GuidelineName"],
                    "Analyte": x["AnalyteName"],
                    "Points": list(map(lambda p: {
                        "DateTime": p["DateTime"],
                        "Value": p["Value"]}, x["StandardValuePoints"]))}, http_data["StandardData"]))}
        return data

    def __httpGet(self, url):
        response = self.session.get(url)
        if 200 == response.status_code:
            return response.json()
        else:
            print("Get data fail with status code" + str(response.status_code))
            raise ConnectionError("Get data fail with status code" + str(response.status_code))

    def __httpPost(self, url, json_data):
        response = self.session.post(url, json=json_data)
        if 200 == response.status_code:
            return response.json()
        else:
            print("Get data fail with status code" + str(response.status_code))
            raise ConnectionError("Get data fail with status code" + str(response.status_code))

File: R/WRD_Client.R
#Compatible with WRD V5.0.2.2
# install packages
library(jsonlite)
library(httr)

authenticate_wrd <- function(wrd_url, username, password){
  loginurl <- paste(wrd_url, "/Account/LogOn", sep="")
  #Set user account data and agent
  pars=list(
    Username=username,
    Password=password
  )
  agent="Mozilla/5.0" #or whatever 
  
  res <- POST(loginurl, body = pars, encode = "form")
}

get_response_data <- function(response){
  if(status_code(response) == 200){
    return(content(response))
  }
  
  stop("Fail to get data from WRD, please try again or contact the EIS team.")
}

get_guideline_list <- function(wrd_url){
  guideline_list_part <- "/API/WaterQualityGuideline"
  request_url <- paste(wrd_url, guideline_list_part, sep="")
  response <- GET(request_url)
  return(get_response_data(response))
}

get_guideline_detail_list <- function(wrd_url){
  guideline_detail_list_part <- "/API/WaterQualityGuideline/Detail"
  request_url <- paste(wrd_url, guideline_detail_list_part, sep="")
  response <- GET(request_url)
  return(get_response_data(response))
}

get_location_list <- function(wrd_url){
  location_list_part <- "/API/WaterQuality/Location/List"
  request_url <- paste(wrd_url, location_list_part, sep="")
  res <- GET(request_url)
  location_content = get_response_data(res)
  data = lapply(location_content, function(x){ 
    return(
      list(LocationIdentifier=x$LocationIdentifier,
           Description=x$LocationDescription,
           Latitude=x$Latitude,
           Longtitude=x$Longtitude,
           LocationType=x$LocationType)
    )
  })
  
  return(data)
  
}

get_location_detail <- function(wrd_url, locationId){
  location_detail_part <- "/API/WaterQuality/Location/Detail/"
  request_url <- paste(wrd_url, location_detail_part, locationId, sep="")
  res <- GET(request_url)
  location_content = get_response_data(res)
  
  location_data = lapply(location_content$Locations, function(x){ 
    return(
      list(LocationName=x$Name,
           LocationType=x$Type,
           Notes=x$Notes)
    )
  })
  
  wq_data = lapply(location_content$WaterQualityDatasets, function(x){ 
    return(
      list(LocationName=x$LocationName,
           Analyte=x$Analyte,
           NumberOfRecords=x$NumberOfRecords,
           FirstRecordDate=x$FirstRecordDate,
           LastRecordDate=x$LastRecordDate)
    )
  })
  
  data = list(Locations=location_data, Analytes=wq_data)
  
  return(data)
}



get_report_data <- function(wrd_url,
                            start_date, end_date, station, analytes, guidelines){
  report_data_url <- paste(wrd_url, "/API/WaterQuality/Graph", sep="")
  
  analyte_data = lapply(analytes, function(x){ return(list(Analyte=x,LocationName=station))})
  guideline_data = lapply(guidelines, function(x){ return(list(GuidelineName=x))})
  
  args=list(
    StartDate=start_date,
    EndDate=end_date,
    StationName=station,
    WaterQualityDatasets=analyte_data,
    WaterQualityStandards=guideline_data
  )
  response <- POST(
    url = report_data_url,
    body = toJSON(args, auto_unbox=TRUE),
    content_type_json(),
    encode = "json"
  )
  
  raw_content = get_response_data(response)
  
  analyte_result_data = lapply(raw_content$AnalytesData, 
                               function(x){ 
                                 return(
                                   list(AnalyteName=x$AnalyteName,
                                        Location=x$StationId,
                                        Unit=x$Unit,
                                        Points=lapply(x$Points, function(p){
                                          return(list(
                                            DateTime=p$DateTime,
                                            Value=p$Value,
                                            DetectionLimit=p$DetectionLimit
                                          ))
                                        }))
                                 )})
  
  guideline_result_data = lapply(raw_content$StandardData, 
                                 function(x){ 
                                   return(
                                     list(GuidelineName=x$GuidelineName,
                                          AnalyteName=x$AnalyteName,
                                          Location=x$StationId,
                                          Unit=x$Unit,
                                          Points=lapply(x$StandardValuePoints, function(p){
                                            return(list(
                                              DateTime=p$DateTime,
                                              Value=p$Value
                                            ))
                                          }))
                                   )})
  
  result <- list(AnalytesData=analyte_result_data,GuidelineData=guideline_result_data)
  return(result)
  
}






